# Test corrcond

```{r, echo=T}
  source("R/corrcond.R")
  library(data.table)
  ld_matrix = as.matrix(read.table("/Volumes/Elements/paintor_runs/final_regions/11_5:72375077-72506696.LD.EUR.old", header=F))
  frequencies = read.table("/Volumes/Elements/paintor_runs/final_regions/11_5:72375077-72506696.EUR", header=F)
  kottgen_meta_analysis = fread("~/Dunedin/Tony/Metanalysis/input_smr.txt", header=T)
  all_af = fread("/Volumes/Elements/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/all_af.txt", header=F)
  freq_af = merge(frequencies,all_af,by.x=3,by.y=1)
  colnames(freq_af) = c("rsid","chr","pos","Z","chr1","pos1","a1","a2","pop1","af")
  freq_af$af = as.numeric(freq_af$af)
  
    idxs_with_betas = which(freq_af$rsid %in% kottgen_meta_analysis$SNP)
  #freq_af = merge(freq_af,kottgen_meta_analysis,by=1)
  ld_matrix= ld_matrix[idxs_with_betas,idxs_with_betas]
  freq_af = freq_af[idxs_with_betas,]
  freq_af = merge(freq_af,kottgen_meta_analysis,by=1)
  freq_af = freq_af[order(freq_af$pos),]
  #freq_af$n = freq_af$TotalSampleSize
  #freq_af$se = freq_af$StdErr
  #freq_af$b = freq_af$Effect
  #stepwise_conditional_run(data_set = freq_af)
  stepwise_conditional_wrapper(data_set = freq_af, ld_matrix=ld_matrix,p_value_threshold = 1e-3,var_y = 1.6421, ld_noise=0.1)
    # Generate hwe_diagonal
  
  
  res_preparation = prep_dataset_common(data_set = freq_af,ld_matrix= ld_matrix,ld_noise=0.1)
  stepwise_results = stepwise_conditional_run(res_preparation = res_preparation,p_value_threshold = 1e-3,var_y = 1.6421)
  all_but_one_df = all_but_one(res_preparation=res_preparation,stepwise_results = stepwise_results ,var_y=1.6421)
  

```

# Alright testing the conditional analysis fuction

```{r, echo=T}
    source("R/corrcond.R")
  #  stepwise_results = stepwise_conditional_run(data_set = freq_af, ld_matrix=ld_matrix,p_value_threshold = 0.0001)
    
```

```{r, echo=T}
  freq_af = read.table("data/snp_summary.txt", header=T)
  ld_matrix = as.matrix(read.table("data/ld_matrix_test.txt", header=F))
  stepwise_results = stepwise_conditional_run(data_set = freq_af, ld_matrix=ld_matrix,p_value_threshold = 1e-4,var_y = 1.6421)
  all_but_one_df = all_but_one(data_set =freq_af, ld_matrix = ld_matrix,stepwise_results = stepwise_results ,var_y=1.6421)

```
# Simulate genotypes and run the all-but-one analysis.

```{r}
require(bindata)

# Parameters of joint distribution
N <- 1000000
p1 <- 0.1
p2 <- 0.5
rho<- -.3
beta = c(0.05, 0,.05,0)
h1 <- rmvbin(N, c(p1,p2), bincorr=(1-rho)*diag(2)+rho)
h2 <- rmvbin(N, c(p1,p2), bincorr=(1-rho)*diag(2)+rho)
h_sq = .6
eta = X %*% beta
y = eta + rnorm(N, 0, sd=sqrt((1-h_sq)))
g1 = (h1[,1] + h2[,1])
g2 = (h1[,2] + h2[,2])
h1 <- rmvbin(N, c(p1,p2), bincorr=(1-rho)*diag(2)+rho)
h2 <- rmvbin(N, c(p1,p2), bincorr=(1-rho)*diag(2)+rho)
g3 = (h1[,1] + h2[,1])
g4= (h1[,2] + h2[,2])

X = cbind(g1,g2,g3,g4)
fitJoint1 = fitJoint

X_small = X[1:1e5,]
y_small = y[1:1e5]
fitJoint = lm(y_small ~ X_small)
summary(fitJoint)
X = apply(X,2, function(x){ 
    af = ( table(x)[3] * 2 +  table(x)[2]) / (2* length(x))
    X_test = x
    X_test[ x == 0] = -2  * af
    X_test[ x == 1] = 1-   2* af
    X_test[ x == 2] = 2-   2* af
  return(X_test)
})



X_test = X[,1] 
X_test[ X_test == 0] = -2  * afs[1]
X_test[ X_test == 1] = 1-   2* afs[1]
X_test[ X_test == 2] = 2-   2* afs[1]
X_mean = apply(X,2,function(x) x - mean(x))


h_sq = .6
eta = X %*% beta
y = eta + rnorm(N, 0, sd=sqrt((1-h_sq)))
X_small = X[1:1e5,]
y_small = y[1:1e5]
fitJoint = lm(y_small ~ X_small)
summary(fitJoint)
fit1 = lm(y_small ~ X_small[,1])
summary(fit1)
fit2 = lm(y_small ~ X_small[,2])
summary(fit2)
fit3 = lm(y_small ~ X_small[,3])
summary(fit3)
fit4 = lm(y_small ~ X_small[,4])
summary(fit4)
ld_matrix = cor(X_small)

afs = apply(X_small,2, function(x){ ( table(x)[1] * 2 +  table(x)[2]) / (2* length(x))})
  


betas = c(coef(fit1)[2],coef(fit2)[2],coef(fit3)[2],coef(fit4)[2])
ses =  c(coef(summary(fit1))[2,2],coef(summary(fit2))[2,2],coef(summary(fit3))[2,2],coef(summary(fit4))[2,2])
input_data = data.frame(rsid=c("rs1","rs2","rs3","rs4"),b=betas,se=ses,n=1e5,af=afs)
#input_data = input_data[c(1,3),]
#input_data = input_data[idx,]
#ld_matrix = ld_matrix[c(1,3),c(1,3)]
df = stepwise_conditional_run(data_set = input_data, ld_matrix = ld_matrix, var_y = var(y_small))

new_data = all_but_one(stepwise_results = df, data_set =input_data, ld_matrix =ld_matrix, var_y=var(y_small))

inside=cor(X_small[,idx])





idx = c(1,3)
fit_two = (lm(y_small ~ X_small[,1] + X_small[,3]))
summary(fit_two)

 hwe_diag =  (2*afs[idx] * ( 1- afs[idx]) * 1e5)

#outside2 = diag(diag(t(X_small)%*%X_small)^1) %*% inside
## We get the same results.
inside=cor(X_small[,idx])

outside = sqrt(diag(hwe_diag)) %*% inside %*% sqrt(diag(hwe_diag))

beta_inv = solve(outside)
#Covariances
new_betas = beta_inv %*% diag(hwe_diag) %*% betas[idx]
new_betas
coef(fit_two)[2:3]

inside=cor(X_small)
inside



inside= diag(diag(t(X_small)%*%X_small)^-0.5) %*% t(X_small)%*% X_small %*% diag(diag(t(X_small)%*%X_small)^-0.5) 
inside




outside = sqrt(diag(diag(t(X_small)%*%X_small))) %*% inside %*% sqrt(diag(diag(t(X_small)%*%X_small)))

beta_inv = solve(outside)
#Covariances
new_betas = beta_inv %*% diag(diag(t(X_small)%*%X_small)) %*% betas
new_betas
coef(fitJoint)[2:5]




?cor







inside= diag(diag(t(X_small[,idx])%*%X_small[,idx])^-0.5) %*% t(X_small[,idx])%*% X_small[,idx] %*% diag(diag(t(X_small[,idx])%*%X_small[,idx])^-0.5) 
inside = cor(X_small[,idx])
outside = sqrt(diag(diag(t(X_small[,idx])%*%X_small[,idx]))) %*% inside %*% sqrt(diag(diag(t(X_small[,idx])%*%X_small[,idx])))
#outside2 = diag(diag(t(X_small)%*%X_small)^1) %*% inside
beta_inv = solve(outside)
#Covariances
new_betas = beta_inv %*% diag(diag(t(X_small[,idx])%*%X_small[,idx])) %*% betas[idx]
new_betas
coef(fit_two)[2:3]

inside= diag(diag(t(X_small[,idx])%*%X_small[,idx])^-0.5) %*% t(X_small[,idx])%*% X_small[,idx] %*% diag(diag(t(X_small[,idx])%*%X_small[,idx])^-0.5) 

outside = sqrt(diag(diag(t(X_small[,idx])%*%X_small[,idx]))) %*% inside %*% sqrt(diag(diag(t(X_small[,idx])%*%X_small[,idx])))
#outside2 = diag(diag(t(X_small)%*%X_small)^1) %*% inside
beta_inv = solve(outside)
#Covariances
new_betas = beta_inv %*% diag(diag(t(X_small[,idx])%*%X_small[,idx])) %*% betas[idx]
new_betas





coef(fit_two)[2:3]
#t((var(y) -t(t(apply(X,2,var))) * t(t(betas^2))) / 1e5)%*% beta_inv
vars = (t(y_small) %*% y_small - t(new_betas) %*% diag(diag(t(X_small[,idx])%*%X_small[,idx])) %*% ((betas[idx]))) / (1e5 - 2)
sqrt(diag(vars[1] * beta_inv))
coef(summary(fit_two))[,2][2:3]




=======
```



# Testing with daner files and chr2_200825237_I_2merged_2_200628237_201293237
```{r, echo=T}
  source("R/symmetrcize.R")
  ld_matrix = as.matrix(read.table("data/chr2_200825237_I_2merged_2_200628237_201293237.ld", header=F))
  ld_matrix = symmetricize(ld_matrix)
  snpdat = read.table("data/chr2_200825237_I_2merged_2_200628237_201293237.ld.snpdat", header=T)
  daner = read.table("data/daner_PGC_SCZ52_0513a_hq2-chr2_200825237_I_2merged_2_200628237_201293237.txt", header=T)
  daner_dat = merge(daner,snpdat, by="SNP")
  daner_dat$b <- log(daner_dat$OR)
  daner_dat$se = daner_dat$SE
  daner_dat$n = daner_dat$sum_n
  colnames(daner_dat)[c(1,20,21)] = c("rsid","af","info")
  daner_dat = daner_dat[order(daner_dat$BP.x),]
  r2 =  r_inverse(daner_dat$b ,0.5,daner_dat$af)
  r(daner_dat$b,phi = 0.5,daner_dat$af)
  gammas = r2[,2]
  
  u = 0.43097855797849721204
  alpha = log(u/(1-u))
  gamma = gammas * exp(alpha) / ((1+exp(alpha))^2)
  linear_ses = daner_dat$se * exp(alpha) / ((1+exp(alpha))^2)
  
  y = 2*(daner_dat$af) * (1 - daner_dat$af) * daner_dat$sum_n
  x = 1/beta^2
  summary(lm(y ~ x - 1))
  plot(daner_dat$b - r(gamma,phi = 0.5, theta = daner_dat$af))
  
  
    #


daner_dat$b 

    res_preparation = prep_dataset_common(data_set = daner_dat,ld_matrix= ld_matrix,ld_noise=0)
  stepwise_results = stepwise_conditional_run(res_preparation = res_preparation,p_value_threshold = 1e-3,var_y = 3.7288262)
  all_but_one_df = all_but_one(res_preparation=res_preparation,stepwise_results = stepwise_results ,var_y=3.7288262)
  
  
  all_but_one_df = all_but_one(data_set =daner_dat, ld_matrix = ld_matrix, stepwise_results = stepwise_resultsd ,var_y=3.7288262)
  
  test = daner_dat[c(515,14,1648),]
  ld_test = ld_matrix[c(515,14,1648),c(515,14,1648)]
  stepwise_results = stepwise_conditional_run(data_set = test, ld_matrix = ld_test, p_value_threshold = 1e-4,  var_y = 3.7288262)  
  
r_inverse= function(gamma,phi,ref_af){
  B =  (0.5*(1-2*phi) * (1 - 2*ref_af) * gamma -1)
  A = - (0.084 + 0.9 * phi * (1-2 * phi) * ref_af * (1-ref_af)) * gamma
  C = gamma
  x1 = (-B+ sqrt( B^2 - 4 * A * C  ))/(2*A)
  x2 = (-B - sqrt( B^2 - 4 * A * C  ))/(2*A)
  return(cbind(x1,x2))
  }
  r =  function(gamma,phi,theta){
    r = gamma/(((phi*(1-phi)) + 0.5*(1-2*phi) * (1 - 2*theta) * gamma) - (0.084 + 0.9 * phi * (1-2 * phi) * theta * (1-theta)) /(phi*(1-phi))* gamma^2)
    r
  }
  
  u = 0.4
alpha = log(u/(1-u))
  0.2610864* exp(alpha) / ((1+exp(alpha))^2)
  0.1215532 * exp(alpha) / ((1+exp(alpha))^2)
  r(0.03,0.5,0.944)
  
 u = 35476 /(46839 + 35476)
```





